// solidity-docgen/dist/site.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildSite = exports.DOC_ITEM_CONTEXT = exports.pageAssigner = void 0;
const path_1 = require("path");
const utils_1 = require("solidity-ast/utils");
const doc_item_1 = require("./doc-item");
const clone_1 = require("./utils/clone");
exports.pageAssigner = {
    single: (_1, _2, { pageExtension: ext }) => 'index' + ext,
    items: (item, _, { pageExtension: ext }) => item.name + ext,
    files: (_, file, { sourcesDir, pageExtension: ext }) => file.absolutePath.startsWith(sourcesDir)
        ? (0, path_1.relative)(sourcesDir, file.absolutePath).replace('.sol', ext)
        : file.absolutePath,
};
exports.DOC_ITEM_CONTEXT = '__item_context';
function buildSite(builds, siteConfig) {
    var _a;
    const assign = typeof siteConfig.pages === 'string' ? exports.pageAssigner[siteConfig.pages] : siteConfig.pages;
    const seen = new Set();
    const items = [];
    const pages = {};
    for (const originalBuild of builds) {
        // Clone because we will mutate in order to add item context.
        const output = (0, clone_1.clone)(originalBuild.output);
        const input = (0, clone_1.clone)(originalBuild.input);
        const deref = (0, utils_1.astDereferencer)(output);
        const build = { input, output, deref };
        for (const { ast: file } of Object.values(build.output.sources)) {
            // Some files may appear in different builds but we only use one.
            if (seen.has(file.src))
                continue;
            seen.add(file.src);
            for (const topLevelItem of file.nodes) {
                if (!(0, doc_item_1.isDocItem)(topLevelItem))
                    continue;
                const page = assign(topLevelItem, file, siteConfig);
                const withContext = Object.assign(topLevelItem, {
                    __item_context: { page, node: topLevelItem, file, build },
                });
                items.push(withContext);
                if (page !== undefined) {
                    ((_a = pages[page]) !== null && _a !== void 0 ? _a : (pages[page] = [])).push(withContext);
                }
                for (const node of (0, utils_1.findAll)(doc_item_1.docItemTypes, topLevelItem)) {
                    if (node === topLevelItem)
                        continue;
                    const contract = topLevelItem.nodeType === 'ContractDefinition' ? topLevelItem : undefined;
                    const __item_context = { page, node: node, contract, file, build };
                    Object.assign(node, { __item_context });
                }
            }
        }
    }
    return {
        items,
        pages: Object.entries(pages).map(([id, items]) => ({ id, items })),
    };
}
exports.buildSite = buildSite;
